File: Baekjoon/17000~17999/[17485]진우의 달 여행 (Large)/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

#define endl "\n"
#define INF 987'654'321

vector<vector<vector<int>>> dp;
vector<vector<int>> matrix;
int N, M, res;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> N >> M;

  dp = vector<vector<vector<int>>>(N + 1, vector<vector<int>>(M + 1, vector<int>(3, 0)));
  matrix = vector<vector<int>>(N + 1, vector<int>(M + 1));
  res = INF;

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++)
      cin >> matrix[i][j];
  }

  for (int j = 0; j < M; j++) {
    for (int k = 0; k < 3; k++)
      dp[0][j][k] = matrix[0][j];
  }

  for (int i = 1; i <= N; i++) {
    for (int j = 0; j < M; j++) {
      if (j == 0) {
        dp[i][j][1] = dp[i - 1][j + 1][0] + matrix[i][j];
        dp[i][j][2] = min(dp[i - 1][j][1], dp[i - 1][j + 1][0]) + matrix[i][j];
      } else if (j == M - 1) {
        dp[i][j][0] = min(dp[i - 1][j][1], dp[i - 1][j - 1][2]) + matrix[i][j];
        dp[i][j][1] = dp[i - 1][j - 1][2] + matrix[i][j];
      } else {
        dp[i][j][0] = min(dp[i - 1][j][1], dp[i - 1][j - 1][2]) + matrix[i][j];
        dp[i][j][1] = min(dp[i - 1][j + 1][0], dp[i - 1][j - 1][2]) + matrix[i][j];
        dp[i][j][2] = min(dp[i - 1][j + 1][0], dp[i - 1][j][1]) + matrix[i][j];
      }
    }
  }

  for (int j = 0; j < M; j++) {
    for (int k = 0; k < 3; k++) {
      if (dp[N][j][k] == 0) continue;
      res = min(res, dp[N][j][k]);
    }
  }

  cout << res << endl;

  return 0;
}

File: Baekjoon/17000~17999/[17485]진우의 달 여행 (Large)/solution.py
N, M = list(map(int, input().split()))
INF = 987654321
res = INF

dp = [[[0 for i in range(3)] for j in range(M + 1)] for k in range(N + 1)]
matrix = []

for i in range(N):
    tmp = list(map(int, input().split()))
    tmp.append(0)
    matrix.append(tmp)
matrix.append([0 for i in range(M + 1)])

for j in range(M):
    for k in range(3):
        dp[0][j][k] = matrix[0][j]

for i in range(1, N + 1):
    for j in range(M):
        if(j == 0):
            dp[i][j][1] = dp[i-1][j+1][0] + matrix[i][j]
            dp[i][j][2] = min(dp[i-1][j+1][0], dp[i-1][j][1]) + matrix[i][j]
        elif(j == M - 1):
            dp[i][j][0] = min(dp[i-1][j][1], dp[i-1][j-1][2]) + matrix[i][j]
            dp[i][j][1] = dp[i-1][j-1][2] + matrix[i][j]
        else:
            dp[i][j][0] = min(dp[i-1][j][1], dp[i-1][j-1][2]) + matrix[i][j]
            dp[i][j][1] = min(dp[i-1][j+1][0], dp[i-1][j-1][2]) + matrix[i][j]
            dp[i][j][2] = min(dp[i-1][j+1][0], dp[i-1][j][1]) + matrix[i][j]

for j in range(M):
    for k in range(3):
        if(dp[N][j][k] == 0):
            continue
        res = min(res, dp[N][j][k])

print(res)
